L1Trigger/Phase2L1ParticleFlow/src/corrector.cc

```diff
@@ -0,0 +1,193 @@
+  if (!filename.empty())
+    init_(filename, "", debug);
+}
+l1tpf::corrector::corrector(const std::string &filename, const std::string &directory, float emfMax, bool debug)
+    : emfMax_(emfMax) {
+  if (!filename.empty())
+    init_(filename, directory, debug);
+}
+
+l1tpf::corrector::corrector(TDirectory *src, float emfMax, bool debug) : emfMax_(emfMax) { init_(src, debug); }
+
+void l1tpf::corrector::init_(const std::string &filename, const std::string &directory, bool debug) {
+  std::string resolvedFileName = filename;
+  if (filename[0] != '/')
+    resolvedFileName = edm::FileInPath(filename).fullPath();
+  TFile *lFile = TFile::Open(resolvedFileName.c_str());
+  if (!lFile || lFile->IsZombie())
+    throw cms::Exception("Configuration", "cannot read file " + filename);
+
+  TDirectory *dir = directory.empty() ? lFile : lFile->GetDirectory(directory.c_str());
+  if (!dir)
+    throw cms::Exception("Configuration", "cannot find directory '" + directory + "' in file " + filename);
+  init_(dir, debug);
+
+  lFile->Close();
+}
+
+void l1tpf::corrector::init_(TDirectory *lFile, bool debug) {
+  TH1 *index = (TH1 *)lFile->Get("INDEX");
+  if (!index)
+    throw cms::Exception("Configuration")
+        << "invalid input file " << lFile->GetPath() << ": INDEX histogram not found.\n";
+  index_.reset((TH1 *)index->Clone());
+  index_->SetDirectory(nullptr);
+
+  is2d_ = index_->InheritsFrom("TH2");
+
+  edm::CPUTimer timer;
+  timer.start();
+  std::unordered_map<std::string, TGraph *> graphs;
+  TKey *key;
+  TIter nextkey(lFile->GetListOfKeys());
+  while ((key = (TKey *)nextkey())) {
+    if (strncmp(key->GetName(), "eta_", 4) == 0) {
+      TGraph *gr = (TGraph *)key->ReadObj();
+      if (!gr->TestBit(TGraph::kIsSortedX))
+        gr->Sort();
+      graphs[key->GetName()] = gr;
+    }
+  }
+
+  neta_ = index_->GetNbinsX();
+  nemf_ = (is2d_ ? index_->GetNbinsY() : 1);
+  corrections_.resize(neta_ * nemf_);
+  std::fill(corrections_.begin(), corrections_.end(), nullptr);
+  char buff[32];
+  int ngraphs = 0;
+  for (unsigned int ieta = 0; ieta < neta_; ++ieta) {
+    for (unsigned int iemf = 0; iemf < nemf_; ++iemf) {
+      if (is2d_) {
+        snprintf(buff, 31, "eta_bin%d_emf_bin%d", ieta + 1, iemf + 1);
+      } else {
+        snprintf(buff, 31, "eta_bin%d", ieta + 1);
+      }
+      TGraph *graph = graphs[buff];
+      if (debug)
+        std::cout << "   eta bin " << ieta << " emf bin " << iemf << " graph " << buff
+                  << (graph ? " <valid>" : " <nil>") << std::endl;
+      if (graph) {
+        ngraphs++;
+        corrections_[ieta * nemf_ + iemf] = (TGraph *)graph->Clone();
+      }
+      if (std::abs(index_->GetXaxis()->GetBinCenter(ieta + 1)) > 3.0) {
+        break;  // no EMF bins beyond eta = 3
+      }
+    }
+  }
+  timer.stop();
+  if (debug)
+    std::cout << "Read " << ngraphs << " graphs from " << lFile->GetPath() << " in " << timer.realTime() << " s"
```

Comment:
cout is not allowed
consider if dedicated timing of this step is necessary for production code

Reply:
ok I can remove it